cluster/manager/pair.go

```diff
@@ -0,0 +1,336 @@
+		err := e.Value.(cluster.ClusterListener).ProcessPairRequest(
+			&c.selfNode,
+			response,
+		)
+		if err != nil {
+			logrus.Errorf("Unable to notify %v on a a cluster remote pair request: %v",
+				e.Value.(cluster.ClusterListener).String(),
+				err,
+			)
+
+			return nil, err
+		}
+	}
+
+	logrus.Infof("Successfully paired with remote cluster %v", request.SourceClusterId)
+
+	return response, nil
+}
+
+func (c *ClusterManager) DeletePair(
+	pairId string,
+) error {
+	if err := pairDelete(pairId); err != nil {
+		return err
+	}
+	logrus.Infof("Successfully deleted pairing with cluster %v", pairId)
+	return nil
+}
+
+func (c *ClusterManager) GetPair(
+	pairId string,
+) (*api.ClusterPairGetResponse, error) {
+	var err error
+	if pairId == "" {
+		pairId, err = getDefaultPairId()
+		if err != nil {
+			return nil, err
+		}
+	}
+	pair, err := pairGet(pairId)
+	if err != nil {
+		return nil, err
+	}
+	return &api.ClusterPairGetResponse{
+		PairInfo: pair,
+	}, nil
+}
+
+func (c *ClusterManager) EnumeratePairs() (*api.ClusterPairsEnumerateResponse, error) {
+	response := &api.ClusterPairsEnumerateResponse{}
+	pairs, err := pairList()
+	if err != nil {
+		return nil, err
+	}
+	response.Pairs = pairs
+	response.DefaultId, err = getDefaultPairId()
+	if err != nil {
+		logrus.Warnf("Error getting default cluster pair: %v", err)
+		return nil, err
+	}
+	return response, nil
+}
+
+func (c *ClusterManager) GetPairToken(
+	reset bool,
+) (*api.ClusterPairTokenGetResponse, error) {
+	kvdb := kvdb.Instance()
+	kvlock, err := kvdb.LockWithID(clusterLockKey, c.config.NodeId)
+	if err != nil {
+		logrus.Errorf("Unable to obtain cluster lock for getting cluster pair token: %v", err)
+		return nil, err
+	}
+	defer kvdb.Unlock(kvlock)
+
+	db, _, err := readClusterInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	if db.PairToken == "" || reset {
+		b := make([]byte, 64)
+		rand.Read(b)
+		db.PairToken = fmt.Sprintf("%x", b)
+
+		_, err = writeClusterInfo(&db)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &api.ClusterPairTokenGetResponse{
+		Token: db.PairToken,
+	}, nil
+}
+
+func pairList() (map[string]*api.ClusterPairInfo, error) {
+	kvdb := kvdb.Instance()
+
+	pairs := make(map[string]*api.ClusterPairInfo)
+	kv, err := kvdb.Enumerate(ClusterPairKey)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range kv {
+		if v.Key == clusterPairDefaultKey {
+			continue
+		}
+		info := &api.ClusterPairInfo{}
+		err = json.Unmarshal(v.Value, &info)
+		if err != nil {
+			return nil, err
+		}
+		pairs[info.Id] = info
+	}
+
+	return pairs, nil
+}
+
+func pairCreate(info *api.ClusterPairInfo, setDefault bool) error {
+	kv := kvdb.Instance()
+	kvp, err := kv.Lock(ClusterPairKey)
+	if err != nil {
+		return err
+	}
+	defer kv.Unlock(kvp)
+
+	key := ClusterPairKey + "/" + info.Id
+	_, err = kv.Create(key, info, 0)
+	if err != nil {
+		return err
+	}
+
+	defaultId, err := getDefaultPairId()
+	if setDefault || err == kvdb.ErrNotFound || defaultId == "" {
```

Comment:
Yes, the first pair created will be marked as the default pair. Not sure I get the question about APIs needing to update it.

Reply:
What i meant was: since the very first one always becomes the default and may not be user's intended behavior,  the need to an API to change the default flag becomes a must.

Reply:
@talakad The CreateRequest also has a flag to set newly created pair as the default.